src/retyped/modulegen.re

```diff
@@ -148,28 +149,39 @@ and function_type_to_bstype ctx f => {
   | Some _ => raise (ModulegenTypeError (not_supported "Type parameters" ctx))
   | None => ()
   };
-  let params =
-    if (List.length formal > 0) {
-      List.map
-        (
-          fun ((_, {typeAnnotation: (loc, t), name, optional}): Ast.Type.Function.Param.t) => (
-            switch name {
-            | Some id => string_of_id id
-            | None => ""
-            },
-            if optional {
-              BsType.Optional (type_to_bstype {...ctx, loc} t)
-            } else {
-              type_to_bstype {...ctx, loc} t
-            }
-          )
-        )
-        formal
+  let argTypes ((_, {typeAnnotation: (loc, t), name, optional}): Ast.Type.Function.Param.t) => (
+    switch name {
+    | Some id => string_of_id id
+    | None => ""
+    },
+    if optional {
+      BsType.Optional (type_to_bstype {...ctx, loc} t)
     } else {
+      type_to_bstype {...ctx, loc} t
+    }
+  );
+  let formalParams = List.map argTypes formal;
+  let restParams =
+    switch rest {
+    | Some (_, {argument}) =>
+      let baseType = argTypes argument;
+      /* rest params cannot be BS-optional */
+      Some (
+        switch baseType {
+        | (id, BsType.Optional t) => (id, t)
+        | t => t
+        }
+      )
+    | None => None
+    };
+  let nominalUnit =
+    if (List.length formalParams == 0 && restParams === None) {
       [("", BsType.Unit)]
+    } else {
+      []
     };
   let return = type_to_bstype {...ctx, loc: rt_loc} rt;
-  BsType.Function params return
+  BsType.Function (List.concat [formalParams, nominalUnit]) restParams return
```

Comment:
Can the `List.conat` be a ternary, with either the formal params or the `nominalUnit`? If I am correct, it should always be one or the other?